packages/db/src/collection.ts

```diff
@@ -430,63 +300,380 @@ export class Collection<T extends object = Record<string, unknown>> {
+                this.derivedDeletes.delete(mutation.key)
+                break
+              case `delete`:
+                this.derivedUpserts.delete(mutation.key)
+                this.derivedDeletes.add(mutation.key)
+                break
+            }
+          }
+        }
+      }
+    }
+
+    // Collect events for changes
+    const events: Array<CollectionEvent<T>> = []
+    this.collectOptimisticChanges(previousState, previousDeletes, events)
+
+    // Emit all events at once
+    this.emitEvents(events)
+  }
+
+  /**
+   * Collect events for optimistic changes
+   */
+  private collectOptimisticChanges(
+    previousUpserts: Map<string, T>,
+    previousDeletes: Set<string>,
+    events: Array<CollectionEvent<T>>
+  ): void {
+    const allKeys = new Set([
+      ...previousUpserts.keys(),
+      ...this.derivedUpserts.keys(),
+      ...previousDeletes,
+      ...this.derivedDeletes,
+    ])
+
+    for (const key of allKeys) {
+      const currentValue = this.get(key)
+      const previousValue = this.getPreviousValue(
+        key,
+        previousUpserts,
+        previousDeletes
+      )
+
+      if (previousValue !== undefined && currentValue === undefined) {
+        events.push({ type: `delete`, key, value: previousValue })
+      } else if (previousValue === undefined && currentValue !== undefined) {
+        events.push({ type: `insert`, key, value: currentValue })
+      } else if (
+        previousValue !== undefined &&
+        currentValue !== undefined &&
+        previousValue !== currentValue
+      ) {
+        events.push({
+          type: `update`,
+          key,
+          value: currentValue,
+          previousValue,
+        })
+      }
+    }
+  }
+
+  /**
+   * Get the previous value for a key given previous optimistic state
+   */
+  private getPreviousValue(
+    key: string,
+    previousUpserts: Map<string, T>,
+    previousDeletes: Set<string>
+  ): T | undefined {
+    if (previousDeletes.has(key)) {
+      return undefined
+    }
+    if (previousUpserts.has(key)) {
+      return previousUpserts.get(key)
+    }
+    return this.syncedData.get(key)
+  }
+
+  /**
+   * Emit multiple events at once to all listeners
+   */
+  private emitEvents(events: Array<CollectionEvent<T>>): void {
+    // Emit to individual event listeners
+    for (const event of events) {
+      this.emitEvent(event)
+    }
+
+    // Convert to ChangeMessage format and emit to subscribeChanges listeners
+    if (events.length > 0) {
+      const changeMessages: Array<ChangeMessage<T>> = events.map((event) => {
+        const changeMessage: ChangeMessage<T> = {
+          type: event.type,
+          key: event.key,
+          value: event.value,
+        }
+
+        if (event.previousValue) {
+          ;(changeMessage as any).previousValue = event.previousValue
+        }
+
+        return changeMessage
+      })
+
+      for (const listener of this.changesBatchListeners) {
+        listener([...changeMessages])
```

Comment:
```suggestion
        listener(changeMessages)
```

changeMessages is already a new array